kroxylicious-test-tools/src/main/java/io/kroxylicious/test/server/MockServer.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.test.server;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+
+import org.apache.kafka.common.message.ApiMessageType;
+import org.apache.kafka.common.protocol.ApiMessage;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import io.netty.bootstrap.ServerBootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.ChannelOption;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.handler.logging.LogLevel;
+import io.netty.handler.logging.LoggingHandler;
+
+import io.kroxylicious.test.codec.DecodedRequestFrame;
+import io.kroxylicious.test.codec.KafkaRequestDecoder;
+import io.kroxylicious.test.codec.KafkaResponseEncoder;
+import io.kroxylicious.test.yaml.YamlMapper;
+
+import static io.kroxylicious.test.codec.KafkaApiMessageConverter.requestConverterFor;
+import static io.kroxylicious.test.codec.KafkaApiMessageConverter.responseConverterFor;
+
+/**
+ * MockServer. Provides a mock kafka broker that can respond with a single
+ * fixed ApiMessage at a time. Intended for per-RPC testing of kroxylicious where
+ * we fire one RPC through the proxy, respond with some known message and then
+ * check the output from the proxy.
+ */
+public final class MockServer implements AutoCloseable {
+    private Channel channel;
+    private EventLoopGroup bossGroup;
+    private EventLoopGroup workerGroup;
+    private final int port;
+    private MockHandler serverHandler;
+
+    private MockServer(JsonNode response, int port) {
+        this.port = start(port, response);
+    }
+
+    private static ApiMessage toApiMessage(JsonNode jsonNode) {
+        if (jsonNode == null) {
+            return null;
+        }
+        var apiMessageTypeNode = jsonNode.get("apiMessageType");
+        var versionNode = jsonNode.get("version");
+        var messageType = ApiMessageType.valueOf(apiMessageTypeNode.asText());
+        var requestConverter = responseConverterFor(messageType);
+        JsonNode responseJson = jsonNode.get("response");
+        return requestConverter.reader().apply(responseJson, versionNode.shortValue());
+    }
+
+    /**
+     * Set the response to be served by the MockServer
+     * @param jsonNode the response
+     */
+    public void setResponse(JsonNode jsonNode) {
```

Comment:
I think this should change to a `Request` class containing the `(apiVersion, ApiKey, ApiMessage)` potentially designed to be easy to decode from json with Jackson, either being directly deserializable or having a `fromJsonNode(node)` static method.

Reply:
Have made this change. The integration tests now know how to convert from YAML to the new Request/Response classes where appropriate.